Add a Timescale x Hyperdrive tutorial (#11674)
cloudflare/cloudflare-docs: content/hyperdrive/tutorials/serverless-timeseries-api-with-timescale/index.md

---
updated: 2023-10-30
difficulty: Beginner
content_type: 📝 Tutorial
pcx_content_type: tutorial
title: Create a serverless, globally distributed time-series API with Timescale
layout: single
---

# Create a serverless, globally distributed time-series API with Timescale

## Overview

> In this tutorial you will learn to build an API on Workers which will ingest and query time-series data stored in Timescale (they make PostgreSQL faster in the cloud). Because we want to use Worker API routes for ingesting data, but we don't want to pay the high PostgreSQL connection penalty, we will be using Hyperdrive to proxy our database connection from the edge.

Reviewer: ```suggestion
In this tutorial you will learn to build an API on Workers which will ingest and query time-series data stored in [Timescale])(https://www.timescale.com/) (they make PostgreSQL faster in the cloud).

We will create and deploy a Worker function that exposes API routes for ingesting data, and use [Hyperdrive](https://developers.cloudflare.com/hyperdrive/) to proxy our database connection from the edge and maintain a connection pool to prevent us having to make a new database connection on every request.
```

You will learn how to:

> - Build and deploy Workers.

Reviewer: ```suggestion
- Build and deploy a Cloudflare Worker.
```

> - Use Worker secrets with Wrangler.

Reviewer: ```suggestion
- Use Worker secrets with the Wrangler CLI 
```

> - Deploy a Timescale service

Reviewer: ```suggestion
- Deploy a Timescale database service
```

> - Integrate Workers, Hyperdrive and Timescale

Reviewer: ```suggestion
- Connect your Worker to your Timescale database service with Hyperdrive
```

Connect to your Timescale database using the Service URL you copied in the last step (it has the password embedded). 

If you're using the default PostgreSQL CLI tool **psql** ([installation instructions here](https://www.timescale.com/blog/how-to-install-psql-on-mac-ubuntu-debian-windows/)) to connect you'd run it like so (substituting your own **Service URL** from the previous step). You could also connect using a graphical tool like PgAdmin.

```bash
$ psql SERVICEURL
```

Once you are connected create your table by pasting the following SQL:

```
CREATE TABLE readings(
  ts timestamptz DEFAULT now() NOT NULL,
  sensor UUID NOT NULL,
  metadata jsonb,
  value numeric NOT NULL
 );

SELECT create_hypertable('readings', 'ts');
```

That's it, Timescale will take care of the rest for you as you ingest and query data.

## 4. Create a database configuration

To create a new Hyperdrive instance you will need:

* Your **SERVICEURL** from Step 2.
* A NAME for your Hyperdrive service. We will use **hyperdrive**

{{<Aside type="note">}}

> Hyperdrive is currently in public beta and is currently free to use for all Workers paid plans.

Reviewer: ```suggestion
Hyperdrive is currently in public beta and is currently free to use for all [Workers paid plans](https://developers.cloudflare.com/workers/platform/pricing/).
```

{{</Aside>}}

> Hyperdrive uses the *create* command with the *--connection-string* argument to pass this information. Run it as follows:

Reviewer: ```suggestion
Hyperdrive uses the `create` command with the `--connection-string` argument to pass this information. Run it as follows:
```

```sh
$ npx wrangler hyperdrive create hyperdrive --connection-string="SERVICEURL"
```

{{<Aside type="note">}}

Hyperdrive will attempt to connect to your database with the provided credentials to verify they are correct before creating a configuration. If you encounter an error when attempting to connect, refer to Hyperdrive's [troubleshooting documentation](/hyperdrive/learning/troubleshooting/) to debug possible causes.

{{</Aside>}}

> This command outputs your Hyperdrive ID, add it to your wrangler.toml by replacing the contents with the following:

Reviewer: ```suggestion
This command outputs your Hyperdrive ID.  You can now bind your Hyperdrive configuration to your Worker in your `wrangler.toml` configuration by replacing the contents with the following:
```

```toml
---
filename: wrangler.toml
---
name = "timescale-api"                                                                                      
main = "src/index.ts"                                                                                       
compatibility_date = "2023-10-30"     

node_compat = true

[[hyperdrive]]
binding = "HYPERDRIVE"
id = "your-id-here"
```

Install the Postgres driver into your Worker project:

```sh
$ npm install pg

```

> Now copy the below Worker code, and replacing the current code in `./src/index.js`.  This code does several things.

Reviewer: ```suggestion
Now copy the below Worker code, and replacing the current code in `./src/index.ts`.  This code does several things.
```

1. uses Hyperdrive to connect to Timescale using the the connection string generated from `env.HYPERDRIVE.connectionString` directly to the driver.
2. creates a `POST` route which accepts an array of JSON readings to insert into Timescale in one transaction
3. creates a `GET` route which will take limit parameter and then return the most recent readings. This could be adapted to filter by ID or by timestamp.

```ts
---

> filename: src/worker.ts

Reviewer: ```suggestion
filename: src/index.ts
```
